LLama/LLamaStatelessExecutor.cs

```diff
@@ -105,13 +106,16 @@ await Task.Run(() => { Context.Eval(tokens, 1); }, cancellationToken)
                     );
                 }
 
+                if(control.ShouldStopGeneration(context, inferenceParams, new int[] { id }))
+                    break;
+
                 // Decode this token into text
                 decoder.Add(id);
                 var decoded = decoder.Read();
                 yield return decoded;
 
                 // Check if any of the antiprompts have been generated
-                if (antiprocessor.Add(decoded))
+                if(control.ShouldStopGeneration(Context, inferenceParams, decoded))
```

Comment:
This block doesn't look quite right, although I'm not sure exactly what the behaviour should be!

If `ShouldStopGeneration(token)` (line 109) returns true then the generation immediately stops. The token is never decoded and is never returned to the user.

If `ShouldStopGeneration(string)` (line 118) returns true then the token has already been converted into a string and has been returned to the user. However, the token is **not** added to `lastTokens`.

I think `yield return decoded` and `lastTokens.Add` should always happen atomically (i.e. either both run, or neither do) -  i.e. should line 118 be moved down below `lastTokens.Add`?

This isn't really a problem in the stateless executor, but I'm concerned if the same thing is a problem in the other executors whree the control flow is more complex?

Reply:
Yes you're right! I've changed the implementation here.